### DISCUSSION TITLE
Forthcoming mas support for bundle IDs in addition to app IDs as arguments

### DISCUSSION THREAD

**Discussion by rgoldberg** [Homebrew/discussions#5801] [Tap maintenance and Homebrew development] [answered]
### TL;DR

To facilitate [mas](https://github.com/mas-cli/mas) accepting bundle IDs everywhere ADAM IDs (fka app IDs) are accepted, I would like to change `mas upgrade` to no longer accept app names in places of ADAM IDs, and to only accept ADAM IDs & bundle IDs like all commands will. Will that cause a problem for any parts of Homebrew?

### Details

I am currently working on supporting [bundle IDs as app ID arguments](https://github.com/mas-cli/mas/issues/541) in addition to ADAM IDs as arguments.

ADAM IDs are `UInt64` IDs for App Store apps; they were formerly known as app IDs in mas.

Henceforth, anything that can uniquely identify an app will be considered an app ID, which includes ADAM IDs & bundle IDs.

Both ID types will be able to be used in the same command line, with, by default, ID arguments that can be parsed as `UInt64` processed as ADAM IDs & all other ID arguments processed as bundle IDs.

`--bundle` (or maybe `--bundle-id`, with a possible `--bundles` or `--bundle-ids` alias, respectively; any recommendations as to which flag name to choose) will process all app ID arguments as bundles IDs, even if they can be parsed as `UInt64`. I doubt that any apps have bundle IDs can be parsed as `UInt64`, but I want to be complete.

Existing command lines that used ADAM IDs will still work, but I wanted to give you a heads up that people might expect to be able to use bundle IDs in mas statements in a `Brewfile` in the future (and to let you provide feedback).

The one wrinkle is that, many years ago, someone allowed `mas upgrade` (& only that subcommand) to accept app names as arguments instead of ADAM IDs. Arguments that could be parsed as `UInt64` are processed as ADAM IDs, but all others are processed as app names (e.g., ADAM ID = 409201541, app name = "Pages", bundle ID = "com.apple.iWork.Pages").

I would like to change `mas upgrade` to only accept ADAM IDs & bundle IDs, not app names. Will that cause a problem for any parts of Homebrew?

Reasons:

- `mas upgrade` should behave the same as all other subcommands.
- App names aren't guaranteed to be unique.
- App names might be able to be changed over time, so they might not be reproducibly useable.
- Users run & see apps based on their folder's display name (which isn't necessarily the same as the actual app name).
- Actual app names aren't even available anymore to mas (without having to add code to fetch & cache the app name from the iTunes Search web API).
  - I would therefore match an "app name" argument against the display name of the app folder (without `.app` extension), but folder names can be changed.
- Once I have finished merging shell completion changes into Apple's Swift Argument Parser, I will be able to generate completion scripts for mas.
  - I will use those scripts to lookup bundle IDs whenever a user tries to complete an app ID argument.
  - The completing argument will be matched against installed ADAM IDs, installed bundle IDs, installed app folder names, and iTunes Search web API app names.
  - The latter 2 will be mapped to inserting the bundle ID on the command line.
  - The iTunes Search web API search obviously should take more time than the 3 prior local matchings; if it's too slow, I will either allow it to be disabled, or just get rid of it.
  - The iTunes Search web API search is obviously more useful for commands like `mas install` than for `mas upgrade`, so it might only be performed where it will be more useful.
- As a stopgap measure, I might have all other commands behave the longterm way, but have `mas uninstall` behave as before unless `--bundle` is used. Then, I can later switch upgrade's default non-UInt64 handling to bundle IDs instead of app names. It might be too convoluted, however, to have upgrade be the sole aberration.

**Comment by rgoldberg** [comment]
@MikeMcQuaid I updated the above proposal. I want to ensure that the forthcoming change will not cause any problems for Homebrew. Please mention anyone else who might need to know about this. I'll answer any concerns, work to devise solutions to any issues people might have with it, etc.

**Reply by MikeMcQuaid** [reply, marked as answer]
For `mas upgrade` we're passing the IDs in this format: https://github.com/MikeMcQuaid/dotfiles/blob/6b16566ba76edbbc27a904beb2fca113a1d6b246/Brewfile#L146-L157

If this still works: great. If this is no longer accepted: you'll break every existing `Brewfile` unless migration code is added to `brew bundle`.

**Reply by rgoldberg** [reply]
@MikeMcQuaid Brew bundle should work with my changes, then.

The integer ID numbers (formerly known in mas as app IDs, but actually ADAM IDs, henceforth to be properly referred to by mas as ADAM IDs) will still be accepted & work everywhere.

I assume that the text names before the integer IDs are just labels (i.e. are not considered as IDs by you), and are not actually used as arguments to `mas`. They wouldn't work as arguments, anyway, except if they were passed to `mas upgrade` (so they wouldn't work for `mas purchase` or `mas install`).

In the future, you'll be able to use text bundle IDs instead of (or mixed with) the integer ADAM IDs.

Sorry if I'm being pedantic, but I really don't want to break brew bundle.

I'll let you know before I release the new version (it isn't ready yet, and I don't have an exact timeframe).

Are there any specific dates or days of the week that are better or worse for you? Specific times of day? Could you test it from mas main before I release? Everything should work fine, but I would like you to test first, and for my release time to be most convenient for you.

**Reply by rgoldberg** [reply]
@MikeMcQuaid Something I forgot to mention: bundle ID arguments are a precursor to supporting iOS & iPadOS apps in mas.

It seems that iOS & iPadOS apps do not store their ADAM IDs locally on a Mac, even though each of them has an ADAM ID on the Apple servers.

If users use bundle IDs to interact with locally installed apps, that will be simpler & more performant than having to lookup ADAM IDs from the iTunes Search web API.

So, mas will be able to support iOS & iPadOS apps via mas sometime in the future. ADAM IDs should work for them, but will be a little slower, and might be more error prone.

It probably makes sense to add an alternate `bundleID:` instead of the `id:` argument in `Brewfile`s.

Those will also be more user friendly than the opaque integer ADAM IDs.

**Reply by MikeMcQuaid** [reply]
> I assume that the text names before the integer IDs are just labels (i.e. are not considered as IDs by you), and are not actually used as arguments to `mas`.

Yup.

> Sorry if I'm being pedantic, but I really don't want to break brew bundle.

Should be able to test this relatively easily by modifying Homebrew to ensure it uses your local build by modifying any/all `mac_app_store*.rb` files in Homebrew/brew.


> Everything should work fine, but I would like you to test first, and for my release time to be most convenient for you.

Would appreciate you testing instead if possible 🙇🏻 



> It probably makes sense to add an alternate `bundleID:` instead of the `id:` argument in `Brewfile`s.

Would be game for this. Perhaps worth a `help wanted` issue in Homebrew/brew here once it's shipped in a stable release.
